src/blob.rs

```diff
@@ -1,86 +1,258 @@
+    path::{Path, PathBuf},
+    pin::Pin,
+};
 
-use crate::record::Record;
+use crate::record::{self, Record};
+
+const BLOB_MAGIC_BYTE: u64 = 0xdeaf_abcd;
+
+type Result<T> = std::result::Result<T, Error>;
 
 /// A `Blob` struct for performing of database,
-#[derive(Debug, Default)]
-pub struct Blob<T> {
+#[derive(Debug)]
+pub struct Blob {
+    index: BTreeMap<Vec<u8>, record::Header>,
     header: Header,
-    records: Vec<T>, // @TODO needs verification, created to yield generic T up
+    file: fs::File,
+    name: FileName,
+    current_offset: u64,
 }
 
-/// # Description
-/// # Examples
-#[derive(Debug, Default)]
-struct Header {
-    magic_byte: u64,
-    version: u32,
-    key_size: u32,
-    flags: u64,
+#[derive(Debug)]
+pub struct WriteFuture {
+    file: fs::File,
+    buf: Option<Vec<u8>>,
+    current_offset: u64,
 }
 
-impl<K> Blob<K>
-where
-    K: Default,
-{
-    /// # Description
-    /// Create new blob from file
-    // @TODO more useful result
-    pub fn from_file<P: AsRef<Path>>(_path: P) -> Result<Self, ()> {
-        // @TODO implement
-        Ok(Self::default())
+impl Future for WriteFuture {
+    type Output = Result<()>;
+
+    fn poll(mut self: Pin<&mut Self>, _waker: &Waker) -> Poll<Self::Output> {
+        let buf = self.buf.take().unwrap();
+        self.file.write_at(&buf, self.current_offset).unwrap();
+        Poll::Ready(Ok(()))
     }
+}
 
-    /// # Description
-    /// Writes given slice to file
-    // @TODO more useful result
-    pub fn write(&mut self, _record: Record<K>) -> Result<(), ()> {
-        // @TODO implement
-        Ok(())
+#[derive(Debug)]
+pub struct ReadFuture {
+    file: fs::File,
+    key: Vec<u8>,
+    loc: Location,
+}
+
+impl Future for ReadFuture {
+    type Output = Result<Record>;
+
+    fn poll(self: Pin<&mut Self>, _waker: &Waker) -> Poll<Self::Output> {
+        let mut buf = vec![0u8; self.loc.size as usize];
+        self.file.read_at(&mut buf, self.loc.offset).unwrap();
+        Poll::Ready(Record::from_raw(&buf).map_err(Error::RecordError))
     }
+}
 
+impl Blob {
     /// # Description
-    /// Reads record data, yields `Ok(Vec<u8>)` if read successful,
-    /// otherwise - `Err`
-    // @TODO more useful result
-    pub fn read(&self, _key: &K) -> Result<Record<K>, ()> {
-        // @TODO implement
-        Ok(Record::new())
+    /// Creates new blob file
+    pub fn open_new(name: FileName) -> Result<Self> {
+        Ok(Self {
+            header: Header::new(),
+            file: fs::OpenOptions::new()
+                .create_new(true)
+                .append(true)
+                .read(true)
+                .open(name.as_path())
+                .map_err(Error::OpenNew)?,
+            name,
+            index: BTreeMap::new(),
+            current_offset: 0,
+        })
+    }
+
+    pub fn boxed(self) -> Box<Self> {
+        Box::new(self)
+    }
+
+    pub fn from_file(path: PathBuf) -> Result<Self> {
+        let file = fs::OpenOptions::new()
```

Comment:
High level logic with stubs functions:
1) open file for blob
2) check header and correctness
3) optional data consistency check for all data
4) open index
5) if open index fails - create a new one from the blob (optional)

Reply:
done